CMR 4385

ygliuvt opened the pull request:
Added associated_collections in variable search response.

ygliuvt commented:
Currently, I have made associated_collections as a field in the variable search response by default. We may want to conditionally include it through a search param flag in the future if it turns out to be preferable.

chris-durbin commented on indexer-app/src/cmr/indexer/services/index_service.clj:
```diff
@@ -350,13 +361,22 @@
         (index-concept context coll-concept parsed-coll-concept options)))))
 
+(defn- index-associated-variable
+  "Index the associated variable conept of the given variable association concept."
```
conept typo

ygliuvt replied:
Done.

chris-durbin commented on system-int-test/test/cmr/system_int_test/search/variable/variable_search_test.clj:
```diff
@@ -305,3 +310,51 @@
+
+          (variables/assert-variable-search
+           [expected-variable1]
+           (variables/search {:variable-name updated-variable1-name})))))))
```
Add another test that deleting a collection causes the collection to be removed from the associated_collections in the variables response.

ygliuvt replied:
Good catch! Added support for cascading delete of variable associations when collections are deleted.

oubiwann commented on indexer-app/src/cmr/indexer/services/index_service.clj:
```diff
@@ -260,6 +262,11 @@
                                context concept :variable-association)]
     (get-elastic-version-with-associations context concept tag-associations variable-associations)))
 
+(defmethod get-elastic-version :variable
```
I'm like the changes like this your making, moving us towards thinking more in terms of a consistent, internal API.

ygliuvt replied:
Thanks.

ygliuvt commented:
As the requirement from MMT changes, this issue has become supporting UMM JSON response format of variable search. I incorporated the UMM JSON change into this PR. So this added support of the UMM JSON search response with the response looks like:
```
curl -g -i "%CMR-ENDPOINT%/variables.umm_json?variable_name=Variable1234&pretty=true"
HTTP/1.1 200 OK
Content-Type: application/vnd.nasa.cmr.umm_results+json;version=1.9; charset=UTF-8
Content-Length: 1177

{
  "hits" : 1,
  "took" : 14,
  "items" : [ {
    "meta" : {
      "revision-id" : 2,
      "deleted" : false,
      "format" : "application/vnd.nasa.cmr.umm+json",
      "provider-id" : "PROV1",
      "native-id" : "var1",
      "concept-id" : "V1200000009-PROV1",
      "revision-date" : "2017-08-14T20:12:43Z",
      "concept-type" : "variable"
    },
    "umm" : {
      "VariableType" : "SCIENCE_VARIABLE",
      "DataType" : "float32",
      "Offset" : 0.0,
      "ScienceKeywords" : [ {
        "Category" : "sk-A",
        "Topic" : "sk-B",
        "Term" : "sk-C"
      } ],
      "Scale" : 1.0,
      "FillValues" : [ {
        "Value" : -9999.0,
        "Type" : "Science"
      } ],
      "Sets" : [ {
        "Name" : "Data_Fields",
        "Type" : "Science",
        "Size" : 2,
        "Index" : 2
      } ],
      "Dimensions" : [ {
        "Name" : "Solution_3_Land",
        "Size" : 3
      } ],
      "Definition" : "Defines the variable",
      "Name" : "Variable1234",
      "Units" : "m",
      "LongName" : "A long UMM-Var name"
    },
    "associations" : {
      "collections" : [ {
        "concept-id" : "C1200000007-PROV1"
      } ]
    }
  } ]
}
```
I have removed associated collections from the JSON response.

abbottry commented on search-app/docs/api.md:
````diff
@@ -3140,6 +3137,65 @@ Content-Length: 292
+```
+curl -g -i "%CMR-ENDPOINT%/variables.umm_json?variable_name=Variable1234&pretty=true"
+HTTP/1.1 200 OK
+Content-Type: application/vnd.nasa.cmr.umm_results+json;version=1.9; charset=UTF-8
````
From a documentation stand point this seems misleading. Versioning isnt implemented yet and I don't recognize this Content-Type, is it accurate?

ygliuvt replied:
Currently, UMM-Var schema versioning by itself is not implemented in CMR, as a result UMM-Var is in the same version as the UMM-C, i.e. at 1.9. This is of course wrong and that's why we have issues to implement UMM-Var versioning. Once that is done, this version will be corrected. This is just what CMR currently returns.

abbottry commented:
Does this endpoint support paging?

ygliuvt commented:
Yes, paging is supported in variable search. I also just sent you an email on the Variable search JSON response change. Basically:

Just letting you know that I have taken your recommendation of changing the field names within Variable JSON response. The Variable search JSON response now looks something like:

```
{
  "hits" : 2,
  "took" : 2,
  "items" : [ {
    "concept_id" : "V1200000007-PROV1",
    "revision_id" : 3,
    "provider_id" : "PROV1",
    "native_id" : "var1",
    "name" : "Variable1",
    "long_name" : "A long UMM-Var name"
  }, {
    "concept_id" : "V1200000008-PROV1",
    "revision_id" : 1,
    "provider_id" : "PROV1",
    "native_id" : "var2",
    "name" : "Variable2",
    "long_name" : "A long UMM-Var name"
  } ]
}
```
To be consistent with this change, we have also decided to change the “variable_name” search parameter to “name”. I have updated the search api to reflect the change and kept the variable_name as a supported search parameter for now so that we don’t break MMT, but you guys should update your variable name search calls to use the “name” parameter.

charleshuang80 commented on search-app/docs/api.md:
```diff
@@ -3086,19 +3081,17 @@ The following parameters are supported when searching for variables.
   * options: pattern, ignore_case
 * concept_id
 * keyword (free text)
   * keyword search is case insensitive and supports wild cards ? and *. There is a limit of 30 wild cards allowed in keyword searches. Within 30 wild cards, there's also limit on the max keyword string length. The longer the max keyword string length, the less number of keywords with wild cards allowed. The following fields are indexed for variable keyword search: variable name, measurement, and  ...
```
is long_name indexed for keyword search also, or did it replace measurement?

ygliuvt replied:
yes. LongName is indexed and used for keyword search. The measurement search parameter is removed from search api doc.

charleshuang80 replied:
ok so maybe the end of this line should be edited to replace measurement with long name

ygliuvt replied:
yes. updated. Thanks.

chris-durbin commented on indexer-app/src/cmr/indexer/services/index_service.clj:
```diff
@@ -374,6 +406,20 @@
+  "Performs the cascade actions of collection deletion,
+  i.e. propagate collection deletion to granules and variables"
+  [context concept-mapping-types concept-id revision-id]
+  (doseq [index (idx-set/get-granule-index-names-for-collection context concept-id)]
```
I've seen this fail on occasion when there are a lot of granules - if an exception is thrown we won't reindex the associated variables. In this case the variable associations would not get updated - what do you think about putting a message on the queue to index the variable instead and doing that part first?

We don't have to hold up the pull request for this, just something to think about.

ygliuvt replied:
I wouldn't worry about this too much. Re-indexing variables is not a crucial task and I expect us to re-index variables periodically like we do with collections anyways once the bulk indexing endpoint is in place.

chris-durbin commented on metadata-db-app/src/cmr/metadata_db/services/concept_service.clj:
```diff
@@ -426,6 +426,30 @@
+                        :latest true})
+        associations (search/find-concepts context search-params)]
+    (doseq [association associations]
+      (save-concept-revision context {:concept-type assoc-type
```
I didn't realize we tombstoned associations - are we cleaning these up with a job?

ygliuvt replied:
yes. We have an old revisions cleanup job.
